src/event/sys/windows.rs

```diff
@@ -0,0 +1,374 @@
+    Result,
+};
+
+const ENABLE_MOUSE_MODE: u32 = 0x0010 | 0x0080 | 0x0008;
+
+lazy_static! {
+    static ref ORIGINAL_CONSOLE_MODE: Mutex<Option<u32>> = Mutex::new(None);
+}
+
+/// Initializes the default console color. It will will be skipped if it has already been initialized.
+fn init_original_console_mode(original_mode: u32) {
+    let mut lock = ORIGINAL_CONSOLE_MODE.lock().unwrap();
+
+    if lock.is_none() {
+        *lock = Some(original_mode);
+    }
+}
+
+/// Returns the original console color, make sure to call `init_console_color` before calling this function. Otherwise this function will panic.
+fn original_console_mode() -> u32 {
+    // safe unwrap, initial console color was set with `init_console_color` in `WinApiColor::new()`
+    ORIGINAL_CONSOLE_MODE
+        .lock()
+        .unwrap()
+        .expect("Original console mode not set")
+}
+
+pub(crate) fn enable_mouse_capture() -> Result<()> {
+    let mode = ConsoleMode::from(Handle::current_in_handle()?);
+    init_original_console_mode(mode.mode()?);
+    mode.set_mode(ENABLE_MOUSE_MODE)?;
+
+    Ok(())
+}
+
+pub(crate) fn disable_mouse_capture() -> Result<()> {
+    let mode = ConsoleMode::from(Handle::current_in_handle()?);
+    mode.set_mode(original_console_mode())?;
+    Ok(())
+}
+
+pub(crate) fn handle_mouse_event(mouse_event: MouseEvent) -> Result<Option<Event>> {
+    if let Ok(Some(event)) = parse_mouse_event_record(&mouse_event) {
+        return Ok(Some(Event::Mouse(event)));
+    }
+    Ok(None)
+}
+
+pub(crate) fn handle_key_event(key_event: KeyEventRecord) -> Result<Option<Event>> {
+    if key_event.key_down {
+        if let Some(event) = parse_key_event_record(&key_event) {
+            return Ok(Some(Event::Key(event)));
+        }
+    }
+
+    Ok(None)
+}
+
+impl From<ControlKeyState> for KeyModifiers {
+    fn from(state: ControlKeyState) -> Self {
+        let shift = (state.has_state(SHIFT_PRESSED));
+        let alt = (state.has_state((LEFT_ALT_PRESSED | RIGHT_ALT_PRESSED)));
+        let control = (state.has_state(LEFT_CTRL_PRESSED | RIGHT_CTRL_PRESSED));
+
+        let mut modifier = KeyModifiers::empty();
+
+        if shift {
+            modifier |= KeyModifiers::SHIFT;
+        }
+        if control {
+            modifier |= KeyModifiers::CONTROL;
+        }
+        if alt {
+            modifier |= KeyModifiers::ALT;
+        }
+
+        modifier
+    }
+}
+
+fn parse_key_event_record(key_event: &KeyEventRecord) -> Option<KeyEvent> {
+    let modifiers = KeyModifiers::from(key_event.control_key_state);
+
+    let key_code = key_event.virtual_key_code as i32;
+
+    let parse_result = match key_code {
+        VK_SHIFT | VK_CONTROL | VK_MENU => None,
+        VK_BACK => Some(KeyCode::Backspace),
+        VK_ESCAPE => Some(KeyCode::Esc),
+        VK_RETURN => Some(KeyCode::Enter),
+        VK_F1..=VK_F24 => Some(KeyCode::F((key_event.virtual_key_code - 111) as u8)),
+        VK_LEFT | VK_UP | VK_RIGHT | VK_DOWN => Some(match key_code {
```

Comment:
I'm a bit puzzle with this ... What is the reason to have them grouped, use nested `match` with `_` & `unreachable!` instead of simple ...

```
VK_LEFT => {},
VK_RIGHT => {},
VK_UP => {},
VK_DOWN => {},
```

... without `unreachable!`?

Reply:
Good one will do, has to do with what we first did here, and slowly cleaning up the parsing.